core/sds-aspect-model-validator/src/main/java/io/openmanufacturing/sds/aspectmodel/shacl/RustLikeFormatter.java

```diff
@@ -0,0 +1,358 @@
+      }
+      
+      highlightToken = extractToken( highlight );
+
+      if ( highlightToken == null ) {
+         // without meaningful position information (line/col), we are not able to provide any additional context/details
+         return message;
+      }
+
+      candidateStatements = rawModel.listStatements()
+            .filterKeep( statement -> spansLine( statement, highlightToken.line() ) )
+            .toList();
+      return formatError( candidateStatements, message );
+   }
+
+   private boolean spansLine( final Statement statement, final int lineNumber ) {
+      return isOnLine( statement.getSubject(), lineNumber )
+            || isOnLine( statement.getPredicate(), lineNumber )
+            || isOnLine( statement.getObject(), lineNumber );
+   }
+
+   private boolean isOnLine( final RDFNode node, final int lineNumber ) {
+      final SmartToken nodePosition = extractToken( node );
+      return nodePosition != null && nodePosition.line() == lineNumber;
+   }
+
+   private String formatError( final List<Statement> statements, final String errorMessage ) {
+      final int prefixWidth = String.valueOf( highlightToken.line() ).length() + 1;
+      final String prefix = " ".repeat( prefixWidth ) + "| ";
+      return errorStyle( String.format( "%s> Error at line %d column %d%n", "-".repeat( prefixWidth ), highlightToken.line(), highlightToken.column() ) )
+            + prefix + System.lineSeparator()
+            + highlightToken.line() + " | " + formatStatements( statements ) + System.lineSeparator()
+            + prefix + " ".repeat( highlightToken.column() ) + errorStyle(
+            "^".repeat( highlightToken.content().length() ) + " " + errorMessage ) + System.lineSeparator()
+            + prefix + System.lineSeparator();
+   }
+
+   private String errorStyle( final String plainText ) {
+      textFormatter.reset();
+      textFormatter.formatError( plainText );
+      return textFormatter.getResult();
+   }
+
+   private String formatStatements( final List<Statement> statements ) {
+      currentColumn = 0;
+      seen.clear();
+      sortSequentially( candidateStatements );
+      knownPositions = candidateStatements.stream()
+            .flatMap( statement -> Stream.of( statement.getSubject(), statement.getPredicate(), statement.getObject() ) )
+            .map( RustLikeFormatter::extractToken )
+            .filter( Objects::nonNull )
+            .filter( token -> token.line() == highlightToken.line() )
+            .map( SmartToken::column )
+            .toList();
+
+      textFormatter.reset();
+      statements.forEach( this::formatStatement );
+      return textFormatter.getResult();
+   }
+
+   // model.listStatements() returns the statements in random order, but we want to format them as they appear in the source document,
+   // so reconstruct the original sequence
+   private void sortSequentially( final List<Statement> statements ) {
+      statements.sort( this::documentOrder );
```

Comment:
This will break if the list happens to be immutable. The method is called multiple times with results of Jena's `iterator.toList()` call, which seems to return mutable lists, but we should not rely on this. I'd change `void sortSequentially` into `List<Statement> sortSequentially` and use for example Guava's ordering: `Ordering.from(comparator).sortedCopy(list)` (documentOrder would have to be turned into a Comparator).
Other solution: Create a new ArrayList from the statements list if (and only if) it's unmodifiable (but I don't know how you would check this), then call `.sort()`.